CURTSYS-3570: DTI agent: close workers on network errors

  enable TCP keepalive mechanism (Linux only)
  optimize code by moving break-stmt up
diff --git a/src/common/dti_connection.c b/src/common/dti_connection.c
index 5a42bb4..9edb901 100644
--- a/src/common/dti_connection.c
+++ b/src/common/dti_connection.c
@@ -93,6 +93,34 @@ IFXOS_PRN_USR_MODULE_CREATE(DTI_CON, IFXOS_PRN_LEVEL_HIGH);
    Connection Handling - Local Function defintions
    ========================================================================= */
 
+#if defined(LINUX)
+/**
+   Enable Socket-Level Option - Keepalive.
+   This is on server side, to detect a broken connection.
+
+\param
+   sock     - socket on which keepalive will be enabled.
+
+\return
+   none
+*/
+DTI_STATIC IFX_void_t DTI_conKeepAliveSet(IFXOS_socket_t sock)
+{
+   IFX_int_t optval = 1;
+   socklen_t optlen = sizeof(optval);
+
+   if (setsockopt(sock, SOL_SOCKET, SO_KEEPALIVE, &optval, optlen) < 0)
+   {
+      IFXOS_PRN_USR_ERR_NL(DTI_CON, IFXOS_PRN_LEVEL_ERR,
+         ("Error: DTI Keepalive set - failed "
+         "(errno (%d) <%s>)"IFXOS_CRLF,
+         errno, strerror(errno)));
+   }
+
+   return;
+}
+#endif
+
 
 /* ============================================================================
    Connection Handling - Global Function defintions
@@ -451,6 +479,9 @@ IFX_int_t DTI_conNewConnectGet(
             return IFX_ERROR;
          }
          DTI_SocketNtoa(&pDtiConNew->sockAddr, addrStr);
+#if defined(LINUX)
+         DTI_conKeepAliveSet(pDtiConNew->nFd);
+#endif
          DTI_PRN_USR_DBG_NL(DTI_CON, DTI_PRN_LEVEL_HIGH,
             ("new connection get - connection from %s, sock = %d."DTI_CRLF,
             addrStr, pDtiConNew->nFd));
diff --git a/src/common/dti_control.c b/src/common/dti_control.c
index a9f350d..4d18c8e 100644
--- a/src/common/dti_control.c
+++ b/src/common/dti_control.c
@@ -2071,7 +2071,7 @@ DTI_STATIC IFX_int_t DTI_Worker(
 
       switch(nIfxRet)
       {
-      case DTI_CON_RECV_STATUS_DONE:
+         case DTI_CON_RECV_STATUS_DONE:
             DTI_CONN_PACKET_IN_INC(pAgentCtx, pWorkerCtx->workerNum);
             if (packetError == DTI_eNoError)
             {
@@ -2118,6 +2118,11 @@ DTI_STATIC IFX_int_t DTI_Worker(
             break;
       }
 
+      if (bConClosed == IFX_TRUE)
+      {
+         break;
+      }
+
       bAutoDevMsgActive = IFX_FALSE;
       for (devIdx = 0; devIdx < DTI_MAX_DEVICE_INTERFACES; devIdx++)
       {
@@ -2146,11 +2151,6 @@ DTI_STATIC IFX_int_t DTI_Worker(
       }
 #     endif
 
-      if (bConClosed == IFX_TRUE)
-      {
-         break;
-      }
-
       if (   (bAutoDevMsgActive         == IFX_TRUE)
           || (bControlAutoCliMsgSupport == IFX_TRUE)
          )
